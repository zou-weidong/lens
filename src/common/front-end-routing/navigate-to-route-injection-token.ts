/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectionToken } from "@ogre-tools/injectable";
import type { Route } from "./front-end-route-injection-token";

type RequiredKeys<T> = Exclude<
  {
    [K in keyof T]: T extends Record<K, T[K]> ? K : never;
  }[keyof T],
  undefined
>;

type ObjectContainingNoRequired<T> = T extends void
  ? never
  : RequiredKeys<T> extends []
  ? any
  : never;

type ObjectContainsNoRequired<T> = T extends ObjectContainingNoRequired<T>
  ? true
  : false;

// TODO: Missing types for:
// - Navigating to route with required parameters, without parameters
type Parameters<TParameters> = ObjectContainsNoRequired<TParameters> extends true
  ? { parameters?: TParameters }
  : { parameters: TParameters };

export type NavigateToRouteOptions<TParameter extends object> = Parameters<TParameter> & BaseNavigateToRouteOptions;

export interface BaseNavigateToRouteOptions {
  query?: Record<string, string>;
  fragment?: string;
  withoutAffectingBackButton?: boolean;
}

export interface NavigateToRoute {
  (route: Route<void>, options?: BaseNavigateToRouteOptions): void;
  <TParameters extends Record<TRequiredKeys, string | number>, TRequiredKeys extends string>(route: Route<TParameters>, opts: NavigateToRouteOptions<TParameters>): void;
  <TParameters extends object>(route: Route<TParameters>, opts?: NavigateToRouteOptions<TParameters>): void;
}

export const navigateToRouteInjectionToken = getInjectionToken<NavigateToRoute>({
  id: "navigate-to-route-injection-token",
});
