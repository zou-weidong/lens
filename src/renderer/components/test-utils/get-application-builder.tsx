/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { LensRendererExtension } from "../../../extensions/lens-renderer-extension";
import rendererExtensionsInjectable from "../../../extensions/renderer-extensions.injectable";
import currentlyInClusterFrameInjectable from "../../routes/currently-in-cluster-frame.injectable";
import type { IObservableArray, ObservableSet } from "mobx";
import { computed, observable, runInAction } from "mobx";
import React from "react";
import { Router } from "react-router";
import subscribeStoresInjectable from "../../kube-watch-api/subscribe-stores.injectable";
import allowedResourcesInjectable from "../../cluster-frame-context/allowed-resources.injectable";
import type { RenderResult } from "@testing-library/react";
import { getByText, fireEvent } from "@testing-library/react";
import type { KubeResource } from "../../../common/rbac";
import type { DiContainer } from "@ogre-tools/injectable";
import clusterStoreInjectable from "../../../common/cluster-store/cluster-store.injectable";
import type { ClusterStore } from "../../../common/cluster-store/cluster-store";
import mainExtensionsInjectable from "../../../extensions/main-extensions.injectable";
import { pipeline } from "@ogre-tools/fp";
import { flatMap, compact, join, get, filter, map, matches, last } from "lodash/fp";
import preferenceNavigationItemsInjectable from "../+preferences/preferences-navigation/preference-navigation-items.injectable";
import navigateToPreferencesInjectable from "../../../common/front-end-routing/routes/preferences/navigate-to-preferences.injectable";
import type { MenuItemOpts } from "../../../main/menu/application-menu-items.injectable";
import applicationMenuItemsInjectable from "../../../main/menu/application-menu-items.injectable";
import type { MenuItemConstructorOptions, MenuItem } from "electron";
import storesAndApisCanBeCreatedInjectable from "../../stores-apis-can-be-created.injectable";
import navigateToHelmChartsInjectable from "../../../common/front-end-routing/routes/cluster/helm/charts/navigate-to-helm-charts.injectable";
import hostedClusterInjectable from "../../cluster-frame-context/hosted-cluster.injectable";
import { ClusterFrameContext } from "../../cluster-frame-context/cluster-frame-context";
import type { Cluster } from "../../../common/cluster/cluster";
import { KubeObjectStore } from "../../../common/k8s-api/kube-object.store";
import clusterFrameContextInjectable from "../../cluster-frame-context/cluster-frame-context.injectable";
import startMainApplicationInjectable from "../../../main/start-main-application/start-main-application.injectable";
import startFrameInjectable from "../../start-frame/start-frame.injectable";
import type { NamespaceStore } from "../+namespaces/store";
import namespaceStoreInjectable from "../+namespaces/store.injectable";
import historyInjectable from "../../navigation/history.injectable";
import type { MinimalTrayMenuItem } from "../../../main/tray/electron-tray/electron-tray.injectable";
import electronTrayInjectable from "../../../main/tray/electron-tray/electron-tray.injectable";
import applicationWindowInjectable from "../../../main/start-main-application/lens-window/application-window/application-window.injectable";
import { getDiForUnitTesting as getRendererDi } from "../../getDiForUnitTesting";
import { getDiForUnitTesting as getMainDi } from "../../../main/getDiForUnitTesting";
import { overrideChannels } from "../../../test-utils/channel-fakes/override-channels";
import trayIconPathsInjectable from "../../../main/tray/tray-icon-path.injectable";
import assert from "assert";
import { openMenu } from "react-select-event";
import userEvent from "@testing-library/user-event";
import lensProxyPortInjectable from "../../../main/lens-proxy/lens-proxy-port.injectable";
import type { Route } from "../../../common/front-end-routing/front-end-route-injection-token";
import type { NavigateToRouteOptions } from "../../../common/front-end-routing/navigate-to-route-injection-token";
import { navigateToRouteInjectionToken } from "../../../common/front-end-routing/navigate-to-route-injection-token";
import type { LensMainExtension } from "../../../extensions/lens-main-extension";
import type { LensExtension } from "../../../extensions/lens-extension";

import extensionInjectable from "../../../extensions/extension-loader/extension/extension.injectable";
import { renderFor } from "./renderFor";
import { RootFrame } from "../../frames/root-frame/root-frame";
import { ClusterFrame } from "../../frames/cluster-frame/cluster-frame";
import hostedClusterIdInjectable from "../../cluster-frame-context/hosted-cluster-id.injectable";

type Callback = (dis: DiContainers) => void | Promise<void>;

type EnableExtensions<T> = (...extensions: T[]) => void;
type DisableExtensions<T> = (...extensions: T[]) => void;

export interface ApplicationBuilder {
  dis: DiContainers;
  setEnvironmentToClusterFrame: () => ApplicationBuilder;

  extensions: {
    renderer: {
      enable: EnableExtensions<LensRendererExtension>;
      disable: DisableExtensions<LensRendererExtension>;
    };

    main: {
      enable: EnableExtensions<LensMainExtension>;
      disable: DisableExtensions<LensMainExtension>;
    };
  };

  allowKubeResource: (resourceName: KubeResource) => ApplicationBuilder;
  beforeApplicationStart: (callback: Callback) => ApplicationBuilder;
  beforeRender: (callback: Callback) => ApplicationBuilder;
  render: () => Promise<RenderResult>;

  tray: {
    click: (id: string) => Promise<void>;
    get: (id: string) => MinimalTrayMenuItem | null;
    getIconPath: () => string;
  };

  applicationMenu: {
    click: (path: string) => void;
  };

  preferences: {
    close: () => void;
    navigate: () => void;
    navigateTo: (route: Route<any>, params: Partial<NavigateToRouteOptions<any>>) => void;
    navigation: {
      click: (id: string) => void;
    };
  };

  helmCharts: {
    navigate: () => void;
  };

  select: {
    openMenu: (id: string) => void;
    selectOption: (menuId: string, labelText: string) => void;
  };
}

interface DiContainers {
  rendererDi: DiContainer;
  mainDi: DiContainer;
}

interface Environment {
  RootComponent: React.ElementType;
  onAllowKubeResource: () => void;
}

export const getApplicationBuilder = () => {
  const mainDi = getMainDi({
    doGeneralOverrides: true,
  });

  const overrideChannelsForWindow = overrideChannels(mainDi);

  const rendererDi = getRendererDi({
    doGeneralOverrides: true,
  });

  overrideChannelsForWindow(rendererDi);

  const dis = { rendererDi, mainDi };

  const clusterStoreStub = {
    provideInitialFromMain: () => {},
    getById: (): null => null,
  } as unknown as ClusterStore;

  rendererDi.override(clusterStoreInjectable, () => clusterStoreStub);
  rendererDi.override(storesAndApisCanBeCreatedInjectable, () => true);
  mainDi.override(clusterStoreInjectable, () => clusterStoreStub);

  const beforeApplicationStartCallbacks: Callback[] = [];
  const beforeRenderCallbacks: Callback[] = [];

  const rendererExtensionsState = observable.set<LensRendererExtension>();
  const mainExtensionsState = observable.set<LensMainExtension>();

  rendererDi.override(subscribeStoresInjectable, () => () => () => {});

  const environments = {
    application: {
      RootComponent: RootFrame,

      onAllowKubeResource: () => {
        throw new Error(
          "Tried to allow kube resource when environment is not cluster frame.",
        );
      },
    } as Environment,

    clusterFrame: {
      RootComponent: ClusterFrame,
      onAllowKubeResource: () => {},
    } as Environment,
  };

  let environment = environments.application;

  rendererDi.override(
    currentlyInClusterFrameInjectable,
    () => environment === environments.clusterFrame,
  );

  rendererDi.override(rendererExtensionsInjectable, () =>
    computed(() => [...rendererExtensionsState]),
  );

  mainDi.override(mainExtensionsInjectable, () =>
    computed(() => [...mainExtensionsState]),
  );

  let trayMenuIconPath: string;

  const traySetMenuItemsMock = jest.fn<any, [MinimalTrayMenuItem[]]>();

  mainDi.override(electronTrayInjectable, () => ({
    start: () => {
      const iconPaths = mainDi.inject(trayIconPathsInjectable);

      trayMenuIconPath = iconPaths.normal;
    },
    stop: () => {},
    setMenuItems: traySetMenuItemsMock,
    setIconPath: (path) => {
      trayMenuIconPath = path;
    },
  }));

  let allowedResourcesState: IObservableArray<KubeResource>;
  let rendered: RenderResult;

  const enableExtensionsFor = <T extends ObservableSet>(
    extensionState: T,
    di: DiContainer,
  ) => {
    const getExtension = (extension: LensExtension) =>
      di.inject(extensionInjectable, extension);

    return (...extensionInstances: LensExtension[]) => {
      const addAndEnableExtensions = () => {
        extensionInstances.forEach(instance => {
          const extension = getExtension(instance);

          extension.register();
        });

        runInAction(() => {
          extensionInstances.forEach((extension) => {
            extensionState.add(extension);
          });
        });
      };

      if (rendered) {
        addAndEnableExtensions();
      } else {
        builder.beforeRender(addAndEnableExtensions);
      }
    };
  };

  const builder: ApplicationBuilder = {
    dis,

    applicationMenu: {
      click: (path: string) => {
        const applicationMenuItems = mainDi.inject(
          applicationMenuItemsInjectable,
        );

        const menuItems = pipeline(
          applicationMenuItems.get(),
          flatMap(toFlatChildren(null)),
          filter((menuItem) => !!menuItem.click),
        );

        const menuItem = menuItems.find((menuItem) => menuItem.path === path);

        if (!menuItem) {
          const availableIds = menuItems.map(get("path")).join('", "');

          throw new Error(
            `Tried to click application menu item with ID "${path}" which does not exist. Available IDs are: "${availableIds}"`,
          );
        }

        menuItem.click?.(
          {
            menu: null as never,
            commandId: 0,
            userAccelerator: null,
            ...menuItem,
          } as MenuItem,
          undefined,
          {},
        );
      },
    },

    tray: {
      get: (id: string) => {
        const lastCall = last(traySetMenuItemsMock.mock.calls);

        assert(lastCall);

        return lastCall[0].find(matches({ id })) ?? null;
      },

      getIconPath: () => trayMenuIconPath,

      click: async (id: string) => {
        const lastCall = last(traySetMenuItemsMock.mock.calls);

        assert(lastCall);

        const trayMenuItems = lastCall[0];

        const menuItem = trayMenuItems.find(matches({ id })) ?? null;

        if (!menuItem) {
          const availableIds = pipeline(
            trayMenuItems,
            filter(item => !!item.click),
            map(item => item.id),
            join(", "),
          );

          throw new Error(`Tried to click tray menu item with ID ${id} which does not exist. Available IDs are: "${availableIds}"`);
        }

        if (!menuItem.enabled) {
          throw new Error(`Tried to click tray menu item with ID ${id} which is disabled.`);
        }

        await menuItem.click?.();
      },
    },

    preferences: {
      close: () => {
        const link = rendered.getByTestId("close-preferences");

        fireEvent.click(link);
      },

      navigate: () => {
        const navigateToPreferences = rendererDi.inject(navigateToPreferencesInjectable);

        navigateToPreferences();
      },

      navigateTo: (route: Route<any>, params: Partial<NavigateToRouteOptions<any>>) => {
        const navigateToRoute = rendererDi.inject(navigateToRouteInjectionToken);

        navigateToRoute(route, params);
      },

      navigation: {
        click: (id: string) => {
          const link = rendered.queryByTestId(`tab-link-for-${id}`);

          if (!link) {
            const preferencesNavigationItems = rendererDi.inject(
              preferenceNavigationItemsInjectable,
            );

            const availableIds = preferencesNavigationItems
              .get()
              .map(get("id"));

            throw new Error(
              `Tried to click navigation item "${id}" which does not exist in preferences. Available IDs are "${availableIds.join('", "')}"`,
            );
          }

          fireEvent.click(link);
        },
      },
    },

    helmCharts: {
      navigate: () => {
        const navigateToHelmCharts = rendererDi.inject(navigateToHelmChartsInjectable);

        navigateToHelmCharts();
      },
    },

    setEnvironmentToClusterFrame: () => {
      environment = environments.clusterFrame;

      allowedResourcesState = observable.array();

      rendererDi.override(allowedResourcesInjectable, () =>
        computed(() => new Set([...allowedResourcesState])),
      );

      const clusterStub = {
        accessibleNamespaces: [],
      } as unknown as Cluster;

      const namespaceStoreStub = {
        contextNamespaces: [],
        items: [],
      } as unknown as NamespaceStore;

      const clusterFrameContextFake = new ClusterFrameContext(
        clusterStub,

        {
          namespaceStore: namespaceStoreStub,
        },
      );

      rendererDi.override(namespaceStoreInjectable, () => namespaceStoreStub);
      rendererDi.override(hostedClusterInjectable, () => clusterStub);
      rendererDi.override(hostedClusterIdInjectable, () => "irrelevant-hosted-cluster-id");
      rendererDi.override(clusterFrameContextInjectable, () => clusterFrameContextFake);

      // Todo: get rid of global state.
      KubeObjectStore.defaultContext.set(clusterFrameContextFake);

      return builder;
    },

    extensions: {
      renderer: {
        enable: enableExtensionsFor(rendererExtensionsState, rendererDi),
        disable: disableExtensionsFor(rendererExtensionsState, rendererDi),
      },

      main: {
        enable: enableExtensionsFor(mainExtensionsState, mainDi),
        disable: disableExtensionsFor(mainExtensionsState, mainDi),
      },
    },

    allowKubeResource: (resourceName) => {
      environment.onAllowKubeResource();

      runInAction(() => {
        allowedResourcesState.push(resourceName);
      });

      return builder;
    },

    beforeApplicationStart(callback: (dis: DiContainers) => void) {
      beforeApplicationStartCallbacks.push(callback);

      return builder;
    },

    beforeRender(callback: (dis: DiContainers) => void) {
      beforeRenderCallbacks.push(callback);

      return builder;
    },

    async render() {
      mainDi.inject(lensProxyPortInjectable).set(42);

      for (const callback of beforeApplicationStartCallbacks) {
        await callback(dis);
      }

      const startMainApplication = mainDi.inject(startMainApplicationInjectable);

      await startMainApplication();

      const applicationWindow = mainDi.inject(applicationWindowInjectable);

      await applicationWindow.start();

      const startFrame = rendererDi.inject(startFrameInjectable);

      await startFrame();

      for (const callback of beforeRenderCallbacks) {
        await callback(dis);
      }

      const history = rendererDi.inject(historyInjectable);

      const render = renderFor(rendererDi);

      rendered = render(
        <Router history={history}>
          <environment.RootComponent />
        </Router>,
      );

      return rendered;
    },

    select: {
      openMenu: (menuId) => {
        const selector = rendered.container.querySelector<HTMLElement>(
          `#${menuId}`,
        );

        assert(selector);

        openMenu(selector);
      },

      selectOption: (menuId, labelText) => {
        const menuOptions = rendered.baseElement.querySelector<HTMLElement>(
          `.${menuId}-options`,
        );

        assert(menuOptions);

        const option = getByText(menuOptions, labelText);

        userEvent.click(option);
      },
    },
  };

  return builder;
};

export type ToFlatChildren = (opts: MenuItemConstructorOptions) => (MenuItemOpts & { path: string })[];

function toFlatChildren(parentId: string | null | undefined): ToFlatChildren {
  return ({ submenu = [], ...menuItem }) => [
    {
      ...menuItem,
      path: pipeline([parentId, menuItem.id], compact, join(".")),
    },
    ...(
      Array.isArray(submenu)
        ? submenu.flatMap(toFlatChildren(menuItem.id))
        : [{
          ...submenu,
          path: pipeline([parentId, menuItem.id], compact, join(".")),
        }]
    ),
  ];
}

const disableExtensionsFor = <T extends ObservableSet>(
  extensionState: T,
  di: DiContainer,
) => {
  const getExtension = (instance: LensExtension) =>
    di.inject(extensionInjectable, instance);

  return (...extensionInstances: LensExtension[]) => {
    extensionInstances.forEach((instance) => {
      const extension = getExtension(instance);

      runInAction(() => {
        extension.deregister();

        extensionState.delete(instance);
      });
    });
  };
};
