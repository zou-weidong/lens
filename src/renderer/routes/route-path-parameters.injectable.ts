/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import type { IComputedValue } from "mobx";
import { computed } from "mobx";
import { matchPath } from "react-router";
import currentPathInjectable from "./current-path.injectable";
import type { Route } from "../../common/front-end-routing/front-end-route-injection-token";
import { getOrInsertWith } from "../utils";

const routePathParametersInjectable = getInjectable({
  id: "route-path-parameters",

  instantiate: (di) => {
    const currentPath = di.inject(currentPathInjectable);
    const pathParametersCache = new Map<Route<unknown>, IComputedValue<Partial<Record<string, string>>>>();

    return <Param>(route: Route<Param>): IComputedValue<Partial<Param>> => (
      getOrInsertWith(pathParametersCache, route, () => computed(() => (
        matchPath(currentPath.get(), {
          path: route.path,
          exact: true,
        })?.params ?? {}
      )))
    );
  },
});

export default routePathParametersInjectable;
