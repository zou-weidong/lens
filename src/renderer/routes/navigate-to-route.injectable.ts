/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import { getInjectable } from "@ogre-tools/injectable";
import { navigateToUrlInjectionToken } from "../../common/front-end-routing/navigate-to-url-injection-token";
import type { NavigateToRoute, NavigateToRouteOptions } from "../../common/front-end-routing/navigate-to-route-injection-token";
import { navigateToRouteInjectionToken } from "../../common/front-end-routing/navigate-to-route-injection-token";
import currentlyInClusterFrameInjectable from "./currently-in-cluster-frame.injectable";
import { buildURL } from "../../common/utils/buildUrl";
import type { Route } from "../../common/front-end-routing/front-end-route-injection-token";

const navigateToRouteInjectable = getInjectable({
  id: "navigate-to-route",

  instantiate: (di) => {
    const navigateToUrl = di.inject(navigateToUrlInjectionToken);

    const currentlyInClusterFrame = di.inject(
      currentlyInClusterFrameInjectable,
    );

    return ((route: Route<object | void>, options: NavigateToRouteOptions<object>) => {
      const url = buildURL(route.path, {
        params: options?.parameters,
        query: options?.query,
        fragment: options?.fragment,
      });

      navigateToUrl(url, {
        ...options,
        forceRootFrame: currentlyInClusterFrame && route.clusterFrame === false,
      });
    }) as NavigateToRoute;
  },

  injectionToken: navigateToRouteInjectionToken,
});

export default navigateToRouteInjectable;
