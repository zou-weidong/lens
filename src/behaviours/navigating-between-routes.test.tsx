/**
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */
import type { DiContainer } from "@ogre-tools/injectable";
import { getInjectable } from "@ogre-tools/injectable";
import React from "react";
import { computed } from "mobx";
import type { RenderResult } from "@testing-library/react";
import { routeSpecificComponentInjectionToken } from "../renderer/routes/route-specific-component-injection-token";
import { observer } from "mobx-react";
import type { Route } from "../common/front-end-routing/front-end-route-injection-token";
import { frontEndRouteInjectionToken } from "../common/front-end-routing/front-end-route-injection-token";
import type { ApplicationBuilder } from "../renderer/components/test-utils/get-application-builder";
import { getApplicationBuilder } from "../renderer/components/test-utils/get-application-builder";
import currentRouteInjectable from "../renderer/routes/current-route.injectable";
import currentPathInjectable from "../renderer/routes/current-path.injectable";
import queryParametersInjectable from "../renderer/routes/query-parameters.injectable";
import { navigateToRouteInjectionToken } from "../common/front-end-routing/navigate-to-route-injection-token";
import routePathParametersInjectable from "../renderer/routes/route-path-parameters.injectable";

describe("navigating between routes", () => {
  let rendererDi: DiContainer;
  let rendered: RenderResult;
  let applicationBuilder: ApplicationBuilder;

  beforeEach(async () => {
    applicationBuilder = getApplicationBuilder();
    rendererDi = applicationBuilder.dis.rendererDi;
  });

  describe("given route without path parameters", () => {
    beforeEach(async () => {
      applicationBuilder.beforeApplicationStart(({ rendererDi }) => {
        rendererDi.register(testRouteWithoutPathParametersInjectable);
        rendererDi.register(testRouteWithoutPathParametersComponentInjectable);
      });

      rendered = await applicationBuilder.render();
    });

    describe("when navigating to route", () => {
      let route: Route;

      beforeEach(() => {
        const navigateToRoute = rendererDi.inject(navigateToRouteInjectionToken);

        route = rendererDi.inject(testRouteWithoutPathParametersInjectable);

        navigateToRoute(route);
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("knows current route", () => {
        const currentRoute = rendererDi.inject(currentRouteInjectable);

        expect(currentRoute.get()).toBe(route);
      });

      it("knows current path", () => {
        const currentPath = rendererDi.inject(currentPathInjectable);

        expect(currentPath.get()).toBe("/some-path");
      });

      it("does not have query parameters", () => {
        const queryParameters = rendererDi.inject(queryParametersInjectable);

        expect(queryParameters.get()).toEqual({});
      });

      it("does not have path parameters", () => {
        const pathParameters = rendererDi.inject(routePathParametersInjectable)(route);

        expect(pathParameters.get()).toEqual({});
      });
    });

    it("when navigating to route with query parameters, knows query parameters", () => {
      const navigateToRoute = rendererDi.inject(navigateToRouteInjectionToken);
      const queryParameters = rendererDi.inject(queryParametersInjectable);

      const route = rendererDi.inject(testRouteWithoutPathParametersInjectable);

      navigateToRoute(route, {
        query: {
          someParameter: "some-value",
          someOtherParameter: "some-other-value",
        },
      });

      expect(queryParameters.get()).toEqual({
        someParameter: "some-value",
        someOtherParameter: "some-other-value",
      });
    });
  });

  describe("given route with optional path parameters", () => {
    beforeEach(async () => {
      applicationBuilder.beforeApplicationStart(({ rendererDi }) => {
        rendererDi.register(routeWithOptionalPathParametersInjectable);
        rendererDi.register(routeWithOptionalPathParametersComponentInjectable);
      });


      rendered = await applicationBuilder.render();
    });

    describe("when navigating to route with path parameters", () => {
      let route: Route<any>;

      beforeEach(() => {
        route = rendererDi.inject(routeWithOptionalPathParametersInjectable);

        const navigateToRoute = rendererDi.inject(navigateToRouteInjectionToken);

        navigateToRoute(route, {
          parameters: {
            someParameter: "some-value",
            someOtherParameter: "some-other-value",
          },
        });
      });

      it("renders", () => {
        expect(rendered.container).toMatchSnapshot();
      });

      it("knows current route", () => {
        const currentRoute = rendererDi.inject(currentRouteInjectable);

        expect(currentRoute.get()).toBe(route);
      });

      it("knows current path", () => {
        const currentPath = rendererDi.inject(currentPathInjectable);

        expect(currentPath.get()).toBe(
          "/some-path/some-value/some-other-value",
        );
      });

      it("knows path parameters", () => {
        const pathParameters = rendererDi.inject(routePathParametersInjectable)(route);

        expect(pathParameters.get()).toEqual({
          someParameter: "some-value",
          someOtherParameter: "some-other-value",
        });
      });
    });

    describe("when navigating to route without path parameters", () => {
      let route: Route<any>;

      beforeEach(() => {
        route = rendererDi.inject(routeWithOptionalPathParametersInjectable);

        const navigateToRoute = rendererDi.inject(navigateToRouteInjectionToken);

        navigateToRoute(route);
      });

      it("knows current route", () => {
        const currentRoute = rendererDi.inject(currentRouteInjectable);

        expect(currentRoute.get()).toBe(route);
      });

      it("knows current path", () => {
        const currentPath = rendererDi.inject(currentPathInjectable);

        expect(currentPath.get()).toBe("/some-path");
      });

      it("knows path parameters", () => {
        const pathParameters = rendererDi.inject(routePathParametersInjectable)(route);

        expect(pathParameters.get()).toEqual({
          someParameter: undefined,
          someOtherParameter: undefined,
        });
      });
    });
  });
});

const testRouteWithoutPathParametersInjectable = getInjectable({
  id: "some-route",
  injectionToken: frontEndRouteInjectionToken,

  instantiate: () => ({
    path: "/some-path",
    clusterFrame: false,
    isEnabled: computed(() => true),
  }),
});

const testRouteWithoutPathParametersComponentInjectable = getInjectable({
  id: "some-route-component",

  instantiate: (di) => ({
    route: di.inject(testRouteWithoutPathParametersInjectable),
    Component: () => <div>Some component</div>,
  }),

  injectionToken: routeSpecificComponentInjectionToken,
});

const routeWithOptionalPathParametersInjectable = getInjectable({
  id: "some-route",
  injectionToken: frontEndRouteInjectionToken,

  instantiate: (): Route<{ someParameter?: string; someOtherParameter?: string }> => ({
    path: "/some-path/:someParameter?/:someOtherParameter?",
    clusterFrame: false,
    isEnabled: computed(() => true),
  }),
});

const routeWithOptionalPathParametersComponentInjectable = getInjectable({
  id: "some-route-component",

  instantiate: (di) => {
    const route = di.inject(routeWithOptionalPathParametersInjectable);
    const pathParameters = di.inject(routePathParametersInjectable)(route);

    return {
      route,

      Component: observer(() => (
        <pre>{JSON.stringify(pathParameters.get(), null, 2)}</pre>
      )),
    };
  },

  injectionToken: routeSpecificComponentInjectionToken,
});

