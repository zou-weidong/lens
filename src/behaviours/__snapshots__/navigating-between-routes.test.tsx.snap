// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigating between routes given route with optional path parameters when navigating to route with path parameters renders 1`] = `
<div>
  <div
    class="ClusterManager"
  >
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
      </div>
      <div
        class="items"
      />
    </div>
    <main>
      <div
        id="lens-views"
      />
      <pre>
        {
  "someParameter": "some-value",
  "someOtherParameter": "some-other-value"
}
      </pre>
    </main>
    <div
      class="HotbarMenu flex column"
    >
      <div
        class="HotbarItems flex column gaps"
      />
      <div
        class="HotbarSelector"
      >
        <i
          class="Icon Icon previous material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="play_arrow"
          >
            play_arrow
          </span>
        </i>
        <div
          class="HotbarIndex"
        >
          <div
            class="badge Badge small clickable"
            id="hotbarIndex"
          >
            0
          </div>
        </div>
        <i
          class="Icon material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="play_arrow"
          >
            play_arrow
          </span>
        </i>
      </div>
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
  </div>
  <div
    class="Notifications flex column align-flex-end"
  />
</div>
`;

exports[`navigating between routes given route without path parameters when navigating to route renders 1`] = `
<div>
  <div
    class="ClusterManager"
  >
    <div
      class="topBar"
    >
      <div
        class="items"
      >
        <i
          class="Icon material interactive focusable"
          data-testid="home-button"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="home"
          >
            home
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-back"
        >
          <span
            class="icon"
            data-icon-name="arrow_back"
          >
            arrow_back
          </span>
        </i>
        <i
          class="Icon material interactive disabled focusable"
          data-testid="history-forward"
        >
          <span
            class="icon"
            data-icon-name="arrow_forward"
          >
            arrow_forward
          </span>
        </i>
      </div>
      <div
        class="items"
      />
    </div>
    <main>
      <div
        id="lens-views"
      />
      <div>
        Some component
      </div>
    </main>
    <div
      class="HotbarMenu flex column"
    >
      <div
        class="HotbarItems flex column gaps"
      />
      <div
        class="HotbarSelector"
      >
        <i
          class="Icon Icon previous material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="play_arrow"
          >
            play_arrow
          </span>
        </i>
        <div
          class="HotbarIndex"
        >
          <div
            class="badge Badge small clickable"
            id="hotbarIndex"
          >
            0
          </div>
        </div>
        <i
          class="Icon material interactive focusable"
          tabindex="0"
        >
          <span
            class="icon"
            data-icon-name="play_arrow"
          >
            play_arrow
          </span>
        </i>
      </div>
    </div>
    <div
      class="StatusBar"
    >
      <div
        class="leftSide"
      />
      <div
        class="rightSide"
      />
    </div>
  </div>
  <div
    class="Notifications flex column align-flex-end"
  />
</div>
`;
